lmdeploy/serve/openai/api_server.py

```diff
@@ -431,8 +440,46 @@ async def completion_stream_generator() -> AsyncGenerator[str, None]:
                     completion_tokens=res.generate_token_len,
                     total_tokens=total_tokens,
                 )
+            delta_message = DeltaMessage(role='assistant', content=res.response)
+            if request.tool_choice != 'none' and VariableInterface.tool_parser is not None:
+                if res.finish_reason == 'stop':
+                    res.finish_reason = 'tool_calls'
+                current_text = current_text + res.response
+                delta_token_ids = res.token_ids if res.token_ids is not None else []
+                current_token_ids = current_token_ids + delta_token_ids
+                tool_delta = VariableInterface.tool_parser.extract_tool_calls_streaming(
+                    previous_text=previous_text,
+                    current_text=current_text,
+                    delta_text=res.response,
+                    previous_token_ids=previous_token_ids,
+                    current_token_ids=current_token_ids,
+                    delta_token_ids=delta_token_ids,
+                    request=request)
+                if tool_delta is not None:
+                    delta_message.tool_calls = tool_delta.tool_calls
+                    delta_message.content = tool_delta.content
+                previous_text = current_text
+                previous_token_ids = current_token_ids
+            elif VariableInterface.enable_reasoning:
```

Comment:
Are `tools_calling` and `enable_reasoning` mutually exclusive?

Reply:
Yes, to align with vllm implementation.